pontoon/messaging/emails.py

```diff
@@ -1,23 +1,246 @@
 from django.utils import timezone
 
+from pontoon.actionlog.models import ActionLog
+from pontoon.base.models import Locale
+from pontoon.insights.models import LocaleInsightsSnapshot
 from pontoon.messaging.utils import html_to_plain_text_with_links
 
 
 log = logging.getLogger(__name__)
 
 
+def _get_monthly_user_data(users, months_ago):
+    month_date = timezone.now() - relativedelta(months=months_ago)
+
+    actions = (
+        ActionLog.objects.filter(
+            performed_by__in=users,
+            created_at__month=month_date.month,
+            created_at__year=month_date.year,
+        )
+        .values("performed_by")
+        .annotate(
+            submitted=Count("id", filter=Q(action_type="translation:created")),
+            reviewed=Count(
+                "id",
+                filter=Q(
+                    action_type__in=["translation:approved", "translation:rejected"]
+                ),
+            ),
+        )
+    )
+
+    return {action["performed_by"]: action for action in actions}
+
+
+def _get_monthly_locale_data(months_ago):
+    month_date = timezone.now() - relativedelta(months=months_ago)
+
+    snapshots = (
+        LocaleInsightsSnapshot.objects.filter(
+            created_at__month=month_date.month,
+            created_at__year=month_date.year,
+        )
+        .values("locale")
+        .annotate(
+            added_source_strings=Sum("new_source_strings"),
+            submitted=Sum(F("human_translations") + F("machinery_translations")),
+            reviewed=Sum(F("peer_approved") + F("rejected")),
+        )
+    )
+
+    return {snapshot["locale"]: snapshot for snapshot in snapshots}
+
+
+def _get_monthly_locale_state(months_ago):
+    month_date = timezone.now() - relativedelta(months=months_ago)
+    last_day = calendar.monthrange(month_date.year, month_date.month)[1]
+
+    snapshots = LocaleInsightsSnapshot.objects.filter(
+        created_at__day=last_day,
+        created_at__month=month_date.month,
+        created_at__year=month_date.year,
+    )
+
+    return {snapshot.locale_id: snapshot for snapshot in snapshots}
+
+
+def _get_monthly_locale_contributors(locales, months_ago):
+    month_date = timezone.now() - relativedelta(months=months_ago)
+
+    actions = ActionLog.objects.filter(
+        performed_by__profile__system_user=False,
+        # Exclude system projects
+        translation__entity__resource__project__system_project=False,
+    )
+
+    # Get contributors that started contributing to the locale in the given month
+    first_contributions = actions.values(
+        "performed_by", "translation__locale"
+    ).annotate(first_contribution_date=Min("created_at"))
+
+    new_locale_contributors = {
+        (entry["translation__locale"], entry["performed_by"])
+        for entry in first_contributions
+        if entry["first_contribution_date"].month == month_date.month
+        and entry["first_contribution_date"].year == month_date.year
```

Comment:
This requires fetching and annotating the entire actionlog, and filtering it in Python. Is there no way of doing the work in the DB?

Reply:
This is not intuitive at all, but: when `values()` and `annotate()` are [combined](https://docs.djangoproject.com/en/5.1/topics/db/aggregation/#values), unique combinations of the fields specified in the `values()` clause are grouped. Only these combinations are then returned and annotated. In our case that means instead of returning 7 million items, we're only returning 7 thousand.

That said, I can filter entries by the given month and year already in the QuerySet, which brings the number of items to about 100.